gradio/oauth.py

```diff
@@ -124,19 +123,50 @@ def _add_mocked_oauth_routes(app: fastapi.FastAPI) -> None:
     @app.get("/login/huggingface")
     async def oauth_login(request: fastapi.Request):
         """Fake endpoint that redirects to HF OAuth page."""
-        return RedirectResponse("/login/callback")
+        # Define target (where to redirect after login)
+        redirect_uri = _generate_redirect_uri(request)
+        return RedirectResponse(
+            "/login/callback?" + urllib.parse.urlencode({"_target_url": redirect_uri})
+        )
 
     @app.get("/login/callback")
     async def oauth_redirect_callback(request: fastapi.Request) -> RedirectResponse:
         """Endpoint that handles the OAuth callback."""
         request.session["oauth_info"] = mocked_oauth_info
-        return RedirectResponse("/")
+        return _redirect_to_target(request)
 
     @app.get("/logout")
     async def oauth_logout(request: fastapi.Request) -> RedirectResponse:
         """Endpoint that logs out the user (e.g. delete cookie session)."""
         request.session.pop("oauth_info", None)
-        return RedirectResponse("/")
+        return _redirect_to_target(request)
+
+
+def _generate_redirect_uri(request: fastapi.Request) -> str:
+    if "_target" in request.query_params:
+        # if `_target` already in query params => respect it
+        target = request.query_params["_target_url"]
+    else:
+        # otherwise => keep query params
+        target = "/?" + urllib.parse.urlencode(request.query_params)
+
+    redirect_uri = request.url_for("oauth_redirect_callback").include_query_params(
+        _target_url=target
+    )
+    redirect_uri = str(redirect_uri)
+    if ".hf.space" in redirect_uri:
```

Comment:
It would better to ensure that ".hf.space" is exactly the domain instead of it just being part of the url. We're already parsing the url so it should be feasible

Reply:
Actually instead of looking at `.hf.space`, what do you think about using `request.url.scheme` to determine whether in http or https mode?

Reply:
if url is localhost or 127.0.0.1 we don't want to redirect to `https`. That's why I update the scheme only for `.hf.space`-based urls.

I tested with 
```py
from starlette.datastructures import URL

url = URL('https://wauplin-gradio-oauth-demo.hf.space/?foo=bar&__theme=system')
# url.netloc == 'wauplin-gradio-oauth-demo.hf.space'
```
we can do `if url.netloc.endswith(".hf.space")` if you'd like but that's all I think

Reply:
Updated in https://github.com/gradio-app/gradio/pull/7034/commits/3c11ea4b0b8cb840dedcb397a1931056e1c27ada to `if url.netloc.endswith(".hf.space")`.